### PR TITLE
IGNITE-16498 Add ability to add handlers to RestModule from other Ignite modules

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-16498

### CODE REVIEW

File: modules/rest-api/src/main/java/org/apache/ignite/rest/ErrorResult.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.ignite.internal.rest;
+package org.apache.ignite.rest;
```
Comment (sashapolo) [inline modules/rest-api/src/main/java/org/apache/ignite/rest/ErrorResult.java:18]: Why did you remove the "internal" package?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestModule.java [hunk 2]
```diff
@@ -17,136 +17,73 @@
...
-    /** Node configuration route. */
-    private static final String NODE_CFG_URL = "/management/v1/configuration/node/";
-
-    /** Cluster configuration route. */
-    private static final String CLUSTER_CFG_URL = "/management/v1/configuration/cluster/";
-
-    /** Path parameter. */
-    private static final String PATH_PARAM = "selector";
-
+public class RestModule implements RestHandlersRegister, IgniteComponent {
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestModule.java:47]: I would suggest to rename this class to `RestComponent` to be consistent with other components

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestModule.java [hunk 3]
```diff
@@ -17,136 +17,73 @@
...
-    /** Presentation of cluster configuration. */
-    private final ConfigurationPresentation<String> clusterCfgPresentation;
-
     /** Netty bootstrap factory. */
     private final NettyBootstrapFactory bootstrapFactory;
 
+    private final SimpleRouter router = new SimpleRouter();
+
     /** Netty channel. */
     private volatile Channel channel;
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/RestModule.java:60]: I would suggest to re-write this logic using the busy lock approach

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/Router.java [hunk 4]
```diff
@@ -17,113 +17,19 @@
...
-    public void addRoute(Route route) {
-        routes.add(route);
-    }
-
+public interface Router {
     /**
      * Finds the route by request.
      *
      * @param req Request.
      * @return Route if founded.
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/Router.java:32]:
```suggestion
     * @return Route if found.
```

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/Router.java [hunk 5]
```diff
@@ -17,113 +17,19 @@
...
      */
-    public Optional<Route> route(HttpRequest req) {
-        return routes.stream().filter(r -> r.match(req)).findFirst();
-    }
+    Optional<Route> route(HttpRequest req);
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/Router.java:34]: we don't use Optionals anywhere, I would suggest to use `@Nullable` instead

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.ignite.rest;
+
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.util.AsciiString;
+
+/**
+ * Allows to configure REST handlers vs routes.
```
Comment (sashapolo) [inline modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java:24]: vs? Did you mean `via`?

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ */
+public interface Routes {
+    /**
+     * Adds the route to router chain.
```
Comment (sashapolo) [inline modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java:28]:
```suggestion
     * Adds the route to the router chain.
```

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * GET query helper.
+     *
+     * @param route      Route.
+     * @param acceptType Accept type.
+     * @param hnd        Actual handler of the request.
+     * @return Router
+     */
+    default Routes get(String route, AsciiString acceptType, RequestHandler hnd) {
```
Comment (sashapolo) [inline modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java:42]: `AsciiString` is netty-specific, do we want to expose that?

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+     * @param hnd        Actual handler of the request.
+     * @return Router
+     */
+    default Routes put(String route, AsciiString acceptType, RequestHandler hnd) {
+        addRoute(new Route(route, HttpMethod.PUT, acceptType.toString(), hnd));
+        return this;
+    }
+
+    /**
+     * Defines a PATCH route.
```
Comment (sashapolo) [inline modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java:73]: This javadoc is inconsistent with other javadocs =)

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/SimpleRouter.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    /** Routes. */
+    private final List<Route> routes;
+
+    /**
+     * Creates a new router with the given list of {@code routes}.
+     *
+     * @param routes Routes.
+     */
+    public SimpleRouter(List<Route> routes) {
+        this.routes = List.copyOf(routes);
```
Comment (sashapolo) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/SimpleRouter.java:50]: 1. This constructor is not used anywhere.
2. `List.copyOf` creates an immutable list, `addRoute` which breaks the contract of this class (e.g. `addRoute` will fail)

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java (hunk 6 above)
Comment (rpuch) [reply modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java]: Rephrased it

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java (hunk 8 above)
Comment (rpuch) [reply modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java]: Removed `AsciiString` usages

File: modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java (hunk 9 above)
Comment (rpuch) [reply modules/rest-api/src/main/java/org/apache/ignite/rest/Routes.java]: Made the other javadocs consistent with this one

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/SimpleRouter.java (hunk 10 above)
Comment (rpuch) [reply modules/rest/src/main/java/org/apache/ignite/internal/rest/routes/SimpleRouter.java]: Removed the constructor, it's easy to add it back if we need it in the future

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/RestModule.java (hunk 3 above)
Comment (rpuch) [reply modules/rest/src/main/java/org/apache/ignite/internal/rest/RestModule.java]: This should probably be implemented in a separate issue.
I filed a ticket about this: https://issues.apache.org/jira/browse/IGNITE-16636

**Review by sashapolo — APPROVED** [review]
